js/src/common/states/ModalManagerState.ts

```diff
@@ -0,0 +1,75 @@
+import Modal from '../components/Modal';
+
+/**
+ * Class used to manage modal state.
+ *
+ * Accessible on the `app` object via `app.modal` property.
+ */
+export default class ModalManagerState {
+  /**
+   * @internal
+   */
+  modal: null | {
```

Comment:
Maybe this should be generic? Then, could we enforce that the attrs correspond to the modal class in question?

Reply:
How would that be done, though? We can't make the class generic as it is only initialised once when the frontend boots in the browser.

We could make the `show` method generic to enforce the second argument (props) matching the Modal class (if we then convert that to TS too), but that wouldn't enforce this property.

Reply:
Looking back at this, I wonder why we didn't have this accept a vnode of type Modal, instead of combining class + attrs dynamically? Then, we could have enforced static prop requirements.

Either way, it's good for now.